gslib/command.py

```diff
@@ -641,6 +641,9 @@ def __init__(self,
       for log_filter in logging_filters:
         self.logger.addFilter(log_filter)
 
+    if self.headers is not None:
+      self.non_metadata_headers = GetNonMetadataHeaders(self.headers)
```

Comment:
What is the default value of this?
Seems like it should have:
```
else:
  self.non_metadata_headers = None
```

Reply:
How did this even work before?

Reply:
Headers always has a non-none value -- it's provided an empty dict by the command initializer. This is more defensive though, so I like it better.

Reply:
You were passing `self.non_metadata_headers` around while it was potentially undeclared, so there probably should have been something like:
```
AttributeError: 'X' object has no attribute 'hi'
```

Reply:
Previously I didn't have non_metadata_headers -- I just made the GetNonMetadataHeaders call right in the CloudApiDelegator call that needed them. Now that we needed to pass them to more places an attribute seemed like a better fit, so it's new. 

The only place where complaints of the sort were raised were in unit tests that didn't instantiate Command through the command runner. This is because of this line in command runner which ensures headers always have a non-null value for user-runnable commands: https://github.com/GoogleCloudPlatform/gsutil/blob/master/gslib/command_runner.py#L354

That's why none of the E2E tests were having problems. As I said, adding the else as you suggested is great and better than what I had, though in practice it's not really necessary outside of tests given how commands are executed.

Reply:
With those unit tests the issue wasn't an attribute error, it was more that GetNonMetadataHeaders was getting called with None, which it didn't expect.

Reply:
I think `AttributeError` should have happened anyways if someone managed a call where `self.headers` was `None`